crates/sui-types/src/messages_checkpoint.rs

```diff
@@ -119,10 +124,149 @@ impl Default for ECMHLiveObjectSetDigest {
+pub struct ObjectCheckpointStates {
+    pub contents: Vec<ObjectCheckpointState>,
+}
+
+impl ObjectCheckpointStates {
+    pub fn new(contents: BTreeMap<ObjectID, Option<ObjectDigest>>) -> Self {
+        Self {
+            contents: contents
+                .into_iter()
+                .map(|(id, digest)| ObjectCheckpointState::new(id, digest))
+                .collect::<Vec<_>>(),
+        }
+    }
+
+    /// Compute the Merkle root of the object checkpoint states.
+    /// Assumes that the object checkpoint states are sorted by object ID.
+    pub fn digest(&self) -> SuiResult<ObjectCheckpointStatesDigest> {
+        let tree = MerkleTree::<Blake2b256>::build_from_unserialized(self.contents.iter())
+            .map_err(|e| SuiError::GenericAuthorityError {
+                error: format!("Failed to build Merkle tree: {}", e),
+            })?;
+        let root = tree.root().bytes();
+        Ok(ObjectCheckpointStatesDigest::from(root))
+    }
+}
+
+#[derive(Debug)]
+pub struct ObjectCheckpointStatesDigest {
+    pub digest: Digest,
+}
+
+impl From<[u8; 32]> for ObjectCheckpointStatesDigest {
+    fn from(digest: [u8; 32]) -> Self {
+        Self {
+            digest: Digest::new(digest),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CheckpointArtifacts {
+    pub latest_object_states: ObjectCheckpointStates,
+    // Add more.. e.g., execution digests, events, etc.
+}
+
+impl CheckpointArtifacts {
+    pub fn digest(&self) -> SuiResult<CheckpointArtifactsDigest> {
+        let latest_object_states_digest = self.latest_object_states.digest()?;
+
+        // When there are more artifacts, we can hash them all into a single digest.
+        Ok(CheckpointArtifactsDigest::from(
+            latest_object_states_digest.digest.into_inner(),
+        ))
+    }
+}
+
+impl From<&[&TransactionEffects]> for CheckpointArtifacts {
+    fn from(effects: &[&TransactionEffects]) -> Self {
+        let mut latest_object_states = BTreeMap::new();
+        for e in effects {
+            for object_change in e.object_changes() {
+                latest_object_states.insert(object_change.id, object_change.output_digest);
+            }
+        }
+
+        Self {
+            latest_object_states: ObjectCheckpointStates::new(latest_object_states),
+        }
+    }
+}
+
+impl From<&[TransactionEffects]> for CheckpointArtifacts {
+    fn from(effects: &[TransactionEffects]) -> Self {
+        let effect_refs: Vec<&TransactionEffects> = effects.iter().collect();
+        Self::from(effect_refs.as_slice())
+    }
+}
+
+impl From<&CheckpointData> for CheckpointArtifacts {
+    fn from(checkpoint_data: &CheckpointData) -> Self {
+        let effects = checkpoint_data
+            .transactions
+            .iter()
+            .map(|tx| &tx.effects)
+            .collect::<Vec<_>>();
+
+        Self::from(effects.as_slice())
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, JsonSchema)]
+pub struct CheckpointArtifactsDigest {
```

Comment:
lets keep all digests in `digests.rs` and follow the existing convention of making a new-type struct, for instance

```
/// A transaction will have a (unique) digest.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, JsonSchema)]
pub struct TransactionDigest(Digest);
```